qcelemental/models/molecule.py

```diff
@@ -753,17 +758,77 @@ def get_molecular_formula(self):
         ClH
 
         """
-        count = collections.Counter(x.title() for x in self.symbols)
+        return self._formula_from_symbols(self.symbols, order)
+
+    @staticmethod
+    def _formula_from_symbols(symbols: List[str], order: str = "alphabetical") -> str:
+        """
+        Returns the molecular formula for a list of symbols.
+
+        Pararmeters
+        -----------
+        symbols: List[str]
+            List of chemical symbols
+        order: str, optional
+            sorting order of the formula. Valid choices are "alphabetical" and "hill".
+
+        Returns
+        -------
+        str
+            The molecular formula
+        """
+
+        supported_orders = ["alphabetical", "hill"]
+        if order.lower() not in supported_orders:
+            raise ValueError(f"Unsupported molecular formula order: {order}. Supported orders are f{supported_orders}.")
+        count = collections.Counter(x.title() for x in symbols)
+        element_order = sorted(count.keys())
+
+        if order.lower() == "hill" and "C" in element_order:
+            if "H" in element_order:
+                element_order.insert(0, element_order.pop(element_order.index("H")))
+            element_order.insert(0, element_order.pop(element_order.index("C")))
 
         ret = []
-        for k in sorted(count.keys()):
+        for k in element_order:
             c = count[k]
             ret.append(k)
             if c > 1:
                 ret.append(str(c))
 
         return "".join(ret)
 
+    @staticmethod
+    def order_molecular_formula(formula: str, order: str = "alphabetical") -> str:
+        """
+        Reorders a molecular formula.
+
+        Pararmeters
+        -----------
+        formula: str
+            A molecular formula
+        order: str, optional
+            sorting order of the formula. Valid choices are "alphabetical" and "hill".
+
+        Returns
+        -------
+        str
+            The molecular formula
+        """
+
+        matches = re.findall("[A-Z][^A-Z]*", formula)
+        count = collections.defaultdict(int)
+        for match in matches:
+            match_n = re.match("(\D+)(\d*)", match)
+            assert match_n
+            if match_n.group(2) == "":
+                n = 1
+            else:
+                n = int(match_n.group(2))
+            count[match_n.group(1)] += n
+        symbols = [k for k, v in count.items() for i in range(v)]
```

Comment:
We should probably `symbol.title()` to make sure things don't slip through like `ch4`.

Reply:
The regex won't match symbols that aren't title case. We don't know if "ch4" should be "CH4" or "Ch4". I've added a check for malformed molecular formulae.